Supporting multiple APIs in a single app

petraszd opened the pull request:
Adds `Api.__init__` parameter `config_prefix`.

Solves issue https://github.com/marshmallow-code/flask-smorest/issues/252 according to [this comment](https://github.com/marshmallow-code/flask-smorest/issues/252#issuecomment-1197287924).

petraszd committed:
Supporting multiple APIs in a single app

Using `Api.__init__` parameter `config_prefix`.

petraszd committed:
Merge branch 'master' into supporting-multiple-apis-in-one-app-via-config-prefix

petraszd committed:
After review: some tests; prefixing with "ext_obj"

petraszd commented on flask_smorest/__init__.py:
```diff
@@ -92,8 +92,7 @@ def register_blueprint(self, blp, *, parameters=None, **options):
         Must be called after app is initialized.
         """
         blp_name = options.get("name", blp.name)
-        if hasattr(blp, "config_prefix"):
```
No point in checking. Later `blp.register_views_in_doc` is called. And it is only available in smorest's blueprints.

lafrech commented on tests/test_api.py:
```diff
@@ -1,15 +1,14 @@
 """Test Api class"""
+import apispec
```
Why move this up there?

petraszd replied:
`A` is a first letter in alphabet. So I guess I've pressed my shortcut for `isort` :). Muscle memory. I will revert it back.

petraszd replied:
OK. Turns out it is what `pre-commit` does on it's own.

lafrech commented on tests/test_api.py:
```diff
@@ -1,15 +1,14 @@
 from flask_smorest.exceptions import MissingAPIParameterError
 
-from .utils import get_schemas, get_responses
+from .utils import get_responses, get_schemas
```
Why change the order?

petraszd replied:
Same here: accidental `isort` run.

lafrech commented on flask_smorest/utils.py:
```diff
@@ -137,3 +137,38 @@ def prepare_response(response, spec, content_type):
+    return result
+
+
+def get_config_key(*, ctx, key):
```
Why kw only args?

petraszd replied:
Main reasoning is for `get_config_value`. There I am forcing kw so it has similar interface as `get_config_value`.

lafrech commented on flask_smorest/utils.py:
```diff
@@ -137,3 +137,38 @@ def prepare_response(response, spec, content_type):
+    return getattr(ctx, "config_prefix", "") + key
+
+
+def get_config_value(*, app, ctx, key, default=None):
```
Why kw only args?

petraszd replied:
I was using `get_config_value` and I myself forgot which of params is suppose to go first? `app` or `ctx`? So to force myself to use correct parameters I've made them kw only.

It is not a big deal. I can easily change it back.

---

Though, I do not know if this function is going to be useful after I will introduce a `current_api` context variable.

lafrech replied:
Sure. I started with a few inline comments before my longer review message.

lafrech reviewed (commented):
Do you think it could be possible to create the config dict at `Api` init then refer to it later?

At `Api.init_app`, get `app[config]`, extract values beginning with prefix, then store as `api.config`.

At `Blueprint` registration, keep a link to api as `self.api`, then get config from `self.api.config`, not `current_app.config`.

Pros:

- Less clumsy "get_config_value" and such code. Resulting code would be easier to read.

Cons:

- Default prefix app would have a bigger dict than other apps since all Flask config values (even those unrelated to flask-smorest) would end up in its `Api` config. Not really an issue, just a discrepancy without consequence. Well, there could be consequences in user code: if someone tries to access a Flask config parameter from the default `Api`, it will work, but then this will fail once they create a prefixed `Api`. Not a blocker, though.

- This freezes config at init. Further changes to `app.config` wouldn't be taken into account. I don't think hot modifying `app.config` was ever meant to be supported and I believed I tried not to do it in the tests. Could this be an issue?

----------------------------------------------------------------------------------------

I just realized that this feature assumes `Blueprint` instances can't be shared across `Api` instances. It wasn't obvious to me at first. Could someone want to have different APIs with common blueprints (routed to different routes, but same code)?

A way around this could be to use something in the route/request mechanism to determine, at request time, which `Api` is targetted. Then pick appropriate config. A context variable could be used to make the config available from the blueprint instance in inner view functions. I hope I'm making sense.

petraszd committed:
Merge branch 'master' into supporting-multiple-apis-in-one-app-via-config-prefix

petraszd committed:
Removing unnecessary import changes

petraszd committed:
Creating `current_api`

petraszd committed:
Removing comment

petraszd committed:
Code Cleanup

petraszd committed:
Tests for etag

petraszd committed:
Docstrings

petraszd committed:
Removing callback approach.

If you are not calling `current_api` you should not pay perf tax
on each request.

petraszd committed:
Merge branch 'master' into supporting-multiple-apis-in-one-app-via-config-prefix

petraszd committed:
Adding test for a partial miss

petraszd committed:
Merge branch 'master' into supporting-multiple-apis-in-one-app-via-config-prefix

petraszd committed:
Prefixed config proxy

petraszd commented on flask_smorest/prefixed_config.py:
```diff
@@ -0,0 +1,16 @@
+from collections.abc import Mapping
+
+
+class PrefixedConfigProxy(Mapping):
```
Maybe I should move it to `utils.py`?

lafrech replied:
Yes. And you cold make it more generic, like `PrefixedMappingProxy`, taking proxied `dict` as input rather than `app`.

Then you'd do

```py
self.config = PrefixedMappingProxy(app.config, self.config_prefix)
```

petraszd replied:
Done.

petraszd committed:
Moving to utils

lafrech commented on flask_smorest/utils.py:
```diff
@@ -137,3 +137,38 @@ def prepare_response(response, spec, content_type):
+
+
+class PrefixedMappingProxy(abc.Mapping):
+    """Mapping to proxy an another mapping using a prefix
```
s/an another/another/

lafrech commented on flask_smorest/utils.py:
```diff
@@ -137,3 +137,38 @@ def prepare_response(response, spec, content_type):
+    .. code-block:: python
+        some_dict = {"foobar_key1": 1, "foobar_key2": 2}
+        assert some_dict["key1"] == 1
+        assert some_dict["key2"] == 2
```
I don't think this code sample really helps. I find it confusing.

An explanation sentence could be fine. Or a code sample with the class instantiation.

petraszd replied:
Oh. I actually forgot to use `PrefixedMappingProxy` and just used a simple dict :). Fixed. I think now it is clearer.

lafrech commented on flask_smorest/etag.py:
```diff
@@ -142,14 +143,26 @@ def check_etag(self, etag_data, etag_schema=None):
+            raise CurrentApiNotAvailableError(
+                "Can't access current API config. "
+                "Make sure that request context is available "
+                "and/or ETag is checked inside flask_smorest views."
```
I don't understand that last line.

petraszd replied:
Removed. I think request context is good enough.

lafrech commented on flask_smorest/globals.py:
```diff
@@ -0,0 +1,20 @@
+
+def _find_current_api():
+    if has_request_context() and "flask-smorest" in current_app.extensions:
+        apis = current_app.extensions["flask-smorest"]["apis"].values()
```
Could we just call current_app and let the Flask exception bubble up, rather than implement our own exception?

Do we need to manage the case where "flask-smorest" is not in app.extensions?

petraszd replied:
Probably not. And if `KeyError` would complain about `"flask-smorest"` it is pretty clear indicator that you need a "flask-smorest".

petraszd replied:
Removed that check.

lafrech commented on flask_smorest/globals.py:
```diff
@@ -0,0 +1,20 @@
+def _find_current_api():
+    if has_request_context() and "flask-smorest" in current_app.extensions:
+        apis = current_app.extensions["flask-smorest"]["apis"].values()
+        blueprints = set(request.blueprints)
```
`request.blueprints` requires werkzeug 2.0.1. setup.py needs to be updated.

lafrech commented on flask_smorest/globals.py:
```diff
@@ -0,0 +1,20 @@
+        for api in (x["ext_obj"] for x in apis):
+            if api._registered_blueprint_names & blueprints:
+                return api
+    return None
```
Iterating and matching on names. This makes me cringe a bit.

I expected this part to be simpler but it was just an intuition. I haven't been digging into this myself, so maybe that's how it is.

Could blueprint name clash occur?

Would it make sense to set a {blp_name: api_obj} mapping in app.extensions["flask-smorest"] when registering blueprints so that the lookup here can be replaced with a simple key access? Or does it sound too hackish?

petraszd replied:
> Would it make sense to set a {blp_name: api_obj} mapping in app.extensions["flask-smorest"] when registering blueprints so that the lookup here can be replaced with a simple key access? Or does it sound too hackish?

I think it is better than a set and for-eaching through all apis. I've updated PR to use `{blp_name: api_obj}` instead. Thanks.

> Could blueprint name clash occur?

No. It raises ValueError.

https://github.com/pallets/flask/blob/main/src/flask/blueprints.py#L383-L391
```python
        if name in app.blueprints:
            ...
            raise ValueError(f"The name '{self_name}' is already registered...")
```

> I expected this part to be simpler but it was just an intuition. I haven't been digging into this myself, so maybe that's how it is.

It is annoying because a blueprint can be registered with multiple APIs. Sometimes even indirectly. For example: you can register a blueprint with one API and then nest it into other blueprint. Which in turn would be registered with some another API.

So I can not mess with a passed blueprint. Which means that I would need to add some kind of `BlueprintProxy` to intercept requests, set `current_api`, and pass them to an original blueprint. Would work I guess. But it would mean that each API request would have an additional function call. And I do not think that `current_api` would be needed 99% of time. So I would rather make getting `current_api` slightly slower than to add an additional frame stack to each API request.

lafrech reviewed (commented):
This looks good.

It is an important change. If I was doing it, I'd abuse `git rebase -i` to rewrite it into two or three meaningful commits, for ease of review and nicer history. If it is too much to ask, I might do it myself at some point before merging.

There are still details I'd like to review, but here's a bunch of comments already.

Thanks.

petraszd committed:
Merge branch 'master' into supporting-multiple-apis-in-one-app-via-config-prefix

petraszd committed:
Code review updates

petraszd committed:
Using dict to speedup `current_api`
